beacon-chain/sync/verify/blob.go

```diff
@@ -0,0 +1,71 @@
+package verify
+
+import (
+	"github.com/pkg/errors"
+	field_params "github.com/prysmaticlabs/prysm/v4/config/fieldparams"
+	"github.com/prysmaticlabs/prysm/v4/consensus-types/blocks"
+	"github.com/prysmaticlabs/prysm/v4/encoding/bytesutil"
+	ethpb "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v4/runtime/version"
+)
+
+var (
+	errBlobVerification          = errors.New("unable to verify blobs")
+	ErrMismatchedBlobBlockRoot   = errors.Wrap(errBlobVerification, "BlockRoot in BlobSidecar does not match the expected root")
+	ErrMismatchedBlobBlockSlot   = errors.Wrap(errBlobVerification, "BlockSlot in BlobSidecar does not match the expected slot")
+	ErrMismatchedBlobCommitments = errors.Wrap(errBlobVerification, "commitments at given slot, root and index do not match")
+	ErrMismatchedProposerIndex   = errors.Wrap(errBlobVerification, "proposer index does not match")
+	ErrIncorrectBlobIndex        = errors.Wrap(errBlobVerification, "incorrect blob index")
+)
+
+// BlobAlignsWithBlock verifies if the blob aligns with the block.
+func BlobAlignsWithBlock(blob *ethpb.BlobSidecar, block blocks.ROBlock) error {
+	if block.Version() < version.Deneb {
+		return nil
+	}
+
+	commits, err := block.Block().Body().BlobKzgCommitments()
+	if err != nil {
+		return err
+	}
+
+	if len(commits) == 0 {
+		return nil
+	}
+
+	if blob.Index >= field_params.MaxBlobsPerBlock {
+		return errors.Wrapf(ErrIncorrectBlobIndex, "blob index %d >= max blobs per block %d", blob.Index, field_params.MaxBlobsPerBlock)
+	}
+
+	// Verify slot
+	blobSlot := blob.Slot
+	blockSlot := block.Block().Slot()
+	if blobSlot != blockSlot {
+		return errors.Wrapf(ErrMismatchedBlobBlockSlot, "slot %d != BlobSidecar.Slot %d", blockSlot, blobSlot)
+	}
+
+	// Verify block and parent roots
+	blockRoot := bytesutil.ToBytes32(blob.BlockRoot)
+	if blockRoot != block.Root() {
```

Comment:
Should we be saving these to variables as well just like you've done above and then only have to call eg. `block.Root()` the one time? Same comment for the other checks below.

Reply:
@saolyn is very cheap and it's only used here and the error message.  I think it's fine to keep it as it is. Might be more messy if we use vars for everything

Reply:
I only mentioned it because it looks like you had started assigning to variables in the first validation:
```
blobSlot := blob.Slot
blockSlot := block.Block().Slot()
if blobSlot != blockSlot {
    blah
}
```

Reply:
@saolyn for ROBlock, the root is a structure variable, if the compiler is smart enough it may even avoid completely a local variable even if you specify it and just put loads to the in-memory position of that root.